doc/man/man1/pcompos.1
.\" RCSid "$Id: pcompos.1,v 1.9 2022/03/18 21:09:16 greg Exp $"
.TH PCOMPOS 1 12/18/97 RADIANCE
.SH NAME
pcompos - composite RADIANCE pictures.
.SH SYNOPSIS
.B pcompos
[
.B \-h
][
.B "\-x xres"
][
.B "\-y yres"
][
.B "\-b r g b"
][
.B "\-lh h"
][
.B \-la
]
[
.B "\-t min1"
][
.B "\+t max1"
][
.B "\-l lab"
][
.B "=SS"
]
.B "pic1 x1 y1 .."
.br
or
.br
.B pcompos
[
.B "\-a ncols"
][
.B "\-s spacing"
][
.B "\-o x0 y0"
][
options
]
.B "pic1 pic2 .."
.SH DESCRIPTION
.I Pcompos
arranges and composites RADIANCE pictures and sends the result to the
standard output.
Each input picture must be accompanied by an anchor point (unless the
.I \-a
option is used, see below).
This anchor point is the usually position of the picture's
left lower corner in the final output, but can be changed
for individual pictures with an 
.I =SS
option, where
.I S
is one of '-', '+' or '0', indicating the minimum, maximum or center
of the image, respectively.
(For example,
.I =+-
would indicate the anchor is relative to the right lower corner, and
.I =-0
would indicate the anchor is relative to the center of the left
edge.)\0
Negative anchor coordinates result in the input being cropped at the origin.
By default, the size of the output picture will be just large enough
to encompass all the input files.
By specifying a smaller dimension using the
.I \-x
and
.I \-y
options, input files can be cropped at the upper boundary.
Specifying a larger dimension produces a border.
The
.I \-b
option specifies a background color to appear wherever input
files do not cover.
The default value is black (0 0 0).
The
.I \-h
option may be used to reduce the information header size, which
can grow disproportionately after multiple runs of
.I pcompos
and/or
.I pcomb(1).
.PP
If input files overlap, later pictures will overwrite earlier ones.
By default, input files are copied unconditionally within the output
boundaries.
The
.I \-t
option specifies a lower threshold intensity under which input pixels
will not be copied to the output.
The
.I \+t
option specifies an upper threshold.
These options are useful for cutting around irregular boundaries in
the input.
.PP
The
.I \-l
option can be used to specify a label for a specific picture, which
will be given a height determined by the
.I \-lh
option (default 24 pixels) and placed in the upper left corner of
the picture.
This label is generated by the program
.I psign(1).
The
.I \-la
option instructs
.I pcompos
to label each picture automatically by its name.
This is particularly useful in conjunction with the
.I \-a
option for producing a catalog of images (see example below).
The
.I \-l
option may still be used to override the default label
for a picture.
.PP
The
.I \-a
option can be used to automatically compute anchor points
that place successive pictures next to each other in
.I ncols
columns.
The ordering will place the first picture in the lower left corner,
the next just to the right of it, and so on for
.I ncols
pictures.
Then, the next row up repeats the pattern until all the input
pictures have been added to the output.
If the pictures are of different size,
.I pcompos
will end up leaving some background areas in the output picture.
There will also be an unfinished row at the top if the number
of pictures is not evenly divided by
.I ncols.
The
.I "\-s N"
option will cause each image to be separated by at least N pixels.
The
.I "\-o x0 y0"
option specifies a nonzero anchor point for the bottom left image.
If a negative value is given for ncols, then the rows are reversed,
so they start from the top rather than the bottom.
This gives a more natural ordering in most cases, but has the
side-effect of moving any labels from the upper-left of each placed
subimage to the lower-left.
Also, the
.I \-o
option is interpreted differently, where the y0 value is number of
pixels from the top (rather than the bottom).
.PP
The standard input can be specified with a hyphen ('-').
A command that produces a RADIANCE picture can be given in place of a file 
by preceeding it with an exclamation point ('!').
.SH EXAMPLE
To put a copyright label at the bottom of a picture:
.IP "" .2i
psign Copyright 1987 | pcompos inp.hdr 0 0 +t .5 - 384 64 > out.hdr
.PP
To make a catalog of images separated by white 10-pixel borders and
ordered four images per row starting from the top-left:
.IP "" .2i
pcompos -la -a -4 -s 10 -b 1 1 1 dog*.hdr > alldogs.hdr
.SH NOTES
Since there is a limit to the number of open files and processes,
large collections of images must be created in stages.
Even if the system limit on open files is large,
.I pcompos
places an artificial limit of 1024 on the number of open files and/or
processes.
.PP
The EXPOSURE= line is ignored on the input except when there is only
a single input picture, or mutliple pictures all have the same
exposure value.
In such cases,
.I pcompos
will echo the exposure line to the output header appropriately,
so that absolute calibration is maintained.
.SH AUTHOR
Greg Ward
.SH "SEE ALSO"
getinfo(1), pcomb(1), pfilt(1), psign(1), rcrop(1), rpict(1)


src/px/pcompos.c
#ifndef lint
static const char	RCSid[] = "$Id: pcompos.c,v 2.39 2021/09/02 16:10:01 greg Exp $";
#endif
/*
 *  pcompos.c - program to composite pictures.
 *
 *     6/30/87
 */

#include  <math.h>

#include "copyright.h"

#include  "rtio.h"
#include  "platform.h"
#include  "paths.h"
#include  "rterror.h"
#include  "color.h"
#include  "resolu.h"

#ifdef getc_unlocked		/* avoid horrendous overhead of flockfile */
#undef getc
#define getc    getc_unlocked
#endif

#define	 MAXFILE	1024

#define  HASMIN		1
#define  HASMAX		2

					/* output picture size */
int  xsiz = 0;
int  ysiz = 0;
					/* input dimensions */
int  xmin = 0;
int  ymin = 0;
int  xmax = 0;
int  ymax = 0;

COLR  bgcolr = BLKCOLR;			/* background color */

int  labelht = 24;			/* label height */

int  checkthresh = 0;			/* check threshold value */

char  StandardInput[] = "<stdin>";
char  Command[] = "<Command>";
char  Label[] = "<Label>";

char  *progname;

struct {
	char  *name;			/* file or command name */
	FILE  *fp;			/* stream pointer */
	int  xres, yres;		/* picture size */
	int  xloc, yloc;		/* anchor point */
	int  flags;			/* HASMIN, HASMAX */
	double  thmin, thmax;		/* thresholds */
} input[MAXFILE];		/* our input files */

int  nfile;			/* number of files */

int  echoheader = 1;
char  ourfmt[LPICFMT+1] = PICFMT;
int  wrongformat = 0;
double	common_expos = 1.;

double	this_expos;

static gethfunc headline;
static void compos(void);
static int cmpcolr(COLR  c1, double lv2);
static FILE * lblopen(char  *s, int  *xp, int  *yp);



static int
headline(			/* print line preceded by a tab */
	char	*s,
	void	*p
)
{
	char  fmt[MAXFMTLEN];

	if (isheadid(s))
		return(0);
	if (isexpos(s))
		this_expos *= exposval(s);
	if (formatval(fmt, s)) {
		if (globmatch(ourfmt, fmt)) {
			wrongformat = 0;
			strcpy(ourfmt, fmt);
		} else
			wrongformat = 1;
	} else if (echoheader) {
		fputc('\t', stdout);
		fputs(s, stdout);
	}
	return(0);
}


int
main(
	int  argc,
	char  *argv[]
)
{
	int  ncolumns = 0;
	int  autolabel = 0;
	int  curcol = 0, x0 = 0, curx = 0, cury = 0, spacing = 0;
	int  xsgn, ysgn;
	char  *thislabel;
	int  an;
	SET_DEFAULT_BINARY();
	SET_FILE_BINARY(stdin);
	SET_FILE_BINARY(stdout);
	progname = argv[0];

	for (an = 1; an < argc && argv[an][0] == '-'; an++)
		switch (argv[an][1]) {
		case 'h':
			echoheader = !echoheader;
			break;
		case 'x':
			xsiz = atoi(argv[++an]);
			break;
		case 'y':
			ysiz = atoi(argv[++an]);
			break;
		case 'b':
			setcolr(bgcolr, atof(argv[an+1]),
					atof(argv[an+2]),
					atof(argv[an+3]));
			an += 3;
			break;
		case 'a':
			ncolumns = atoi(argv[++an]);
			break;
		case 's':
			spacing = atoi(argv[++an]);
			break;
		case 'o':
			curx = x0 = atoi(argv[++an]);
			cury = atoi(argv[++an]);
			break;
		case 'l':
			switch (argv[an][2]) {
			case 'a':
				autolabel++;
				break;
			case 'h':
				labelht = atoi(argv[++an]);
				break;
			case '\0':
				goto dofiles;
			default:
				goto userr;
			}
			break;
		case '\0':
		case 't':
			goto dofiles;
		default:
			goto userr;
		}
dofiles:
	newheader("RADIANCE", stdout);
	fputnow(stdout);
	for (nfile = 0; an < argc; nfile++) {
		if (nfile >= MAXFILE)
			goto toomany;
		thislabel = NULL;
		input[nfile].flags = 0;
		xsgn = ysgn = '-';
		while (an < argc && (argv[an][0] == '-' || argv[an][0] == '+'
				|| argv[an][0] == '=')) {
			switch (argv[an][1]) {
			case 't':
				checkthresh = 1;
				if (argv[an][0] == '-') {
					input[nfile].flags |= HASMIN;
					input[nfile].thmin = atof(argv[an+1]);
				} else if (argv[an][0] == '+') {
					input[nfile].flags |= HASMAX;
					input[nfile].thmax = atof(argv[an+1]);
				} else
					goto userr;
				an++;
				break;
			case 'l':
				if (strcmp(argv[an], "-l"))
					goto userr;
				thislabel = argv[++an];
				break;
			case '+':
			case '-':
			case '0':
				if (argv[an][0] != '=')
					goto userr;
				xsgn = argv[an][1];
				ysgn = argv[an][2];
				if (ysgn != '+' && ysgn != '-' && ysgn != '0')
					goto userr;
				break;
			case '\0':
				if (argv[an][0] == '-')
					goto getfile;
				goto userr;
			default:
				goto userr;
			}
			an++;
		}
getfile:
		if (argc-an < (ncolumns ? 1 : 3))
			goto userr;
		if (autolabel && thislabel == NULL)
			thislabel = argv[an];
		if (!strcmp(argv[an], "-")) {
			input[nfile].name = StandardInput;
			input[nfile].fp = stdin;
		} else {
			if (argv[an][0] == '!') {
				input[nfile].name = Command;
				input[nfile].fp = popen(argv[an]+1, "r");
			} else {
				input[nfile].name = argv[an];
				input[nfile].fp = fopen(argv[an], "r");
			}
			if (input[nfile].fp == NULL) {
				perror(argv[an]);
				quit(1);
			}
		}
		an++;
						/* get header */
		if (echoheader)
			printf("%s:\n", input[nfile].name);
		this_expos = 1;
		getheader(input[nfile].fp, headline, NULL);
		if (!nfile)
			common_expos = this_expos;
		else if (common_expos > 0 &&
				fabs(this_expos/common_expos - 1.) > 0.02)
			common_expos = 0;
		if (wrongformat) {
			fprintf(stderr, "%s: incompatible input format\n",
					input[nfile].name);
			quit(1);
		}
						/* get picture size */
		if (fgetresolu(&input[nfile].xres, &input[nfile].yres,
				input[nfile].fp) < 0) {
			fprintf(stderr, "%s: bad picture size\n",
					input[nfile].name);
			quit(1);
		}
		if (ncolumns) {
			if (curcol >= abs(ncolumns)) {
				cury = ymax + spacing;
				curx = x0;
				curcol = 0;
			}
			input[nfile].xloc = curx;
			input[nfile].yloc = cury;
			curx += input[nfile].xres + spacing;
			curcol++;
		} else {
			input[nfile].xloc = atoi(argv[an++]);
			if (xsgn == '+')
				input[nfile].xloc -= input[nfile].xres;
			else if (xsgn == '0')
				input[nfile].xloc -= input[nfile].xres/2;
			input[nfile].yloc = atoi(argv[an++]);
			if (ysgn == '+')
				input[nfile].yloc -= input[nfile].yres;
			else if (ysgn == '0')
				input[nfile].yloc -= input[nfile].yres/2;
		}
		if (input[nfile].xloc < xmin)
			xmin = input[nfile].xloc;
		if (input[nfile].yloc < ymin)
			ymin = input[nfile].yloc;
		if (input[nfile].xloc+input[nfile].xres > xmax)
			xmax = input[nfile].xloc+input[nfile].xres;
		if (input[nfile].yloc+input[nfile].yres > ymax)
			ymax = input[nfile].yloc+input[nfile].yres;
		if (thislabel != NULL) {
			if (++nfile >= MAXFILE)
				goto toomany;
			input[nfile].name = Label;
			input[nfile].flags = 0;
			input[nfile].xres = input[nfile-1].xres;
			input[nfile].yres = labelht;
			if ((input[nfile].fp = lblopen(thislabel,
					&input[nfile].xres,
					&input[nfile].yres)) == NULL)
				goto labelerr;
			input[nfile].xloc = input[nfile-1].xloc;
			input[nfile].yloc = input[nfile-1].yloc +
					input[nfile-1].yres-input[nfile].yres;
		}
	}
	if (xsiz <= 0)
		xsiz = xmax;
	else if (xsiz > xmax)
		xmax = xsiz;
	if (ysiz <= 0)
		ysiz = ymax;
	else if (ysiz > ymax)
		ymax = ysiz;
	if (ncolumns < 0) {		/* reverse rows if requested */
		int	i = nfile;
		while (i--)
			input[i].yloc = ymax - input[i].yres - input[i].yloc;
	}
					/* add new header info. */
	printargs(argc, argv, stdout);
	if (common_expos > 0)		/* print exposure if shared */
		fputexpos(common_expos, stdout);
	if (strcmp(ourfmt, PICFMT))
		fputformat(ourfmt, stdout);	/* print format if known */
	putchar('\n');
	fprtresolu(xsiz, ysiz, stdout);

	compos();
	
	quit(0);
userr:
	fprintf(stderr,
	"Usage: %s [-h][-x xr][-y yr][-b r g b][-a n][-s p][-o x0 y0][-la][-lh h] ",
			progname);
	fprintf(stderr, "[-t min1][+t max1][-l lab][=SS] pic1 x1 y1 ..\n");
	quit(1);
toomany:
	fprintf(stderr, "%s: only %d files and labels allowed\n",
			progname, MAXFILE);
	quit(1);
labelerr:
	fprintf(stderr, "%s: error opening label\n", progname);
	quit(1);
	return 1; /* pro forma return */
}


static void
compos(void)				/* composite pictures */
{
	COLR  *scanin, *scanout;
	int  y;
	register int  x, i;

	scanin = (COLR *)malloc((xmax-xmin)*sizeof(COLR));
	if (scanin == NULL)
		goto memerr;
	scanin -= xmin;
	if (checkthresh) {
		scanout = (COLR *)malloc(xsiz*sizeof(COLR));
		if (scanout == NULL)
			goto memerr;
	} else
		scanout = scanin;
	for (y = ymax-1; y >= 0; y--) {
		for (x = 0; x < xsiz; x++)
			copycolr(scanout[x], bgcolr);
		for (i = 0; i < nfile; i++) {
			if (input[i].yloc > y ||
					input[i].yloc+input[i].yres <= y)
				continue;
			if (freadcolrs(scanin+input[i].xloc,
					input[i].xres, input[i].fp) < 0) {
				fprintf(stderr, "%s: read error (y==%d)\n",
						input[i].name,
						y-input[i].yloc);
				quit(1);
			}
			if (y >= ysiz)
				continue;
			if (checkthresh) {
				x = input[i].xloc+input[i].xres;
				if (x > xsiz)
					x = xsiz;
				for (x--; x >= 0 && x >= input[i].xloc; x--) {
					if (input[i].flags & HASMIN &&
					cmpcolr(scanin[x], input[i].thmin) <= 0)
						continue;
					if (input[i].flags & HASMAX &&
					cmpcolr(scanin[x], input[i].thmax) >= 0)
						continue;
					copycolr(scanout[x], scanin[x]);
				}
			}
		}
		if (y >= ysiz)
			continue;
		if (fwritecolrs(scanout, xsiz, stdout) < 0) {
			perror(progname);
			quit(1);
		}
	}
					/* read remainders from streams */
	for (i = 0; i < nfile; i++)
		if (input[i].name[0] == '<')
			while (getc(input[i].fp) != EOF)
				;
	return;
memerr:
	perror(progname);
	quit(1);
}


static int
cmpcolr(			/* compare COLR to luminance */
	register COLR  c1,
	double lv2
)
{
	double	lv1 = .0;
	
	if (c1[EXP])
		lv1 = ldexp((double)normbright(c1), (int)c1[EXP]-(COLXS+8));
	if (lv1 < lv2) return(-1);
	if (lv1 > lv2) return(1);
	return(0);
}


static FILE *
lblopen(		/* open pipe to label generator */
	char  *s,
	int  *xp,
	int  *yp
)
{
	char  com[PATH_MAX];
	FILE  *fp;

	sprintf(com, "psign -s -.15 -a 2 -x %d -y %d '%.90s'", *xp, *yp, s);
	if ((fp = popen(com, "r")) == NULL)
		return(NULL);
	if (checkheader(fp, COLRFMT, NULL) < 0)
		goto err;
	if (fgetresolu(xp, yp, fp) < 0)
		goto err;
	return(fp);
err:
	pclose(fp);
	return(NULL);
}


void
quit(code)		/* exit gracefully */
int  code;
{
	register int  i;
				/* close input files */
	for (i = 0; i < nfile; i++)
		if (input[i].name == Command || input[i].name == Label)
			pclose(input[i].fp);
		else
			fclose(input[i].fp);
	exit(code);
}
